### ISSUE TITLE
ActionMenu focus behavior

### ISSUE THREAD

**Issue by randall-krauskopf** [primer/view_components#2065]
# Context
We are trying to implement hotkeys and custom focus behavior based on key and click events inside the action menu for the tasklist element. There is some default behavior that is preventing us from implementing that. 

## Desired Behavior
We would like to be able to programmatically toggle the focus of an input based on the user selecting an item from the action menu via a mouse click, pressing enter/return, or a `keydown` event driven hotkey.

<details>
  <summary>Example Item click/keyboard working</summary>

https://github.com/primer/view_components/assets/104226843/e287d279-ef20-4273-86b1-17c82c58e067

</details>

### Keyboard and click based focus bug
<details>
  <summary>Keyboard Focus Bug</summary>

https://github.com/primer/view_components/assets/104226843/4504f720-9cd4-4416-9477-b6d8fa1a0f04

</details>

We were able to implement a [temporary timeout fix](https://github.com/github/github/pull/272079/files#diff-9001bfb28bbac746e708f1fc171f96bcfac349b6ecc94adb7ec80a7da167cf98R803-R808) thanks to the suggestions from @keithamus.

### Hotkey Focus

When focused in the action menu, I would like to be able to press the `Option + r` key event to programatically trigger a focus to the list item text area input. Currently the focus goes back to the action menu item button, [setting a timeout similar](https://github.com/github/github/pull/272079/files#diff-9001bfb28bbac746e708f1fc171f96bcfac349b6ecc94adb7ec80a7da167cf98R803-R808) to how we were able to fix it for keyboard and click events is not working the same for some reason.

<details>
  <summary>Hotkey Focus Bug</summary>

https://github.com/primer/view_components/assets/104226843/1171a590-1c33-4ed9-8a76-e15503072531

</details>

<details>
  <summary>Example of behavior in Primer Lookup</summary>

https://github.com/primer/view_components/assets/104226843/61f5a1d2-a05e-4bd7-a2c8-d8bb8185b4e4

</details>

## Misc
* [Slack discussions around click focus](https://github.slack.com/archives/CGJTTJ738/p1684530094467799)
* [Slack discussions around keyboard focus](https://github.slack.com/archives/CGJTTJ738/p1686163810170539)
* [Current related PR](https://github.com/github/github/pull/272079)
  * [Relevant click function](https://github.com/github/github/pull/272079/files#diff-9001bfb28bbac746e708f1fc171f96bcfac349b6ecc94adb7ec80a7da167cf98R769-R777) that is triggered by the [data-action in the context menu](https://github.com/github/github/pull/272079/files#diff-de4f5a59b23f179fbf3b3beae6d93ad80efa280fe577b5a412a51399c9bfaa01R293)

**Comment by dombesz** [comment]
We are also affected by the "Keyboard and click based focus bug". @randall-krauskopf would it be possible to share the [`temporary timeout fix`](https://github.com/github/github/pull/272079/files#diff-9001bfb28bbac746e708f1fc171f96bcfac349b6ecc94adb7ec80a7da167cf98R803-R808) fix publicly?

Thanks!

**Comment by lesliecdubs** [comment]
Thanks for filing! 

@TylerJDev curious for your input on the request above. Do you concur with the suggested change for desired behavior, or should we bring this to Office Hours or Design for deeper discussion?

**Comment by TylerJDev** [comment]
I think this is a fair ask. There should be the ability to point focus elsewhere after activating a menu item. I'll look into this to see if I can figure out why the programmatic focus isn't being respected.

@keithamus might also have some context here too.

**Comment by keithamus** [comment]
I think the problem is because ActionMenu attempts to return focus to its invoking button when the menu closes. The menu close steps should instead be:

1. Close the menu
3. Wait one micro task 
4. If the active element is the document, focus the invoking element.

**Comment by HDinger** [comment]
> We are also affected by the "Keyboard and click based focus bug". This bug also exists when a `Primer::Alpha::Dialog` is rendered with an icon button that opens the dialog. The [icon button](https://github.com/primer/view_components/blob/97f03393977c9c82141eb5255fa3fd619a95e935/app/components/primer/beta/icon_button.rb#L10) requires an `aria-label` to be present all the time and unfortunately there is no easy way to circumvent this.
> Dialog icon button label bug demo
> 
> @randall-krauskopf would it be possible to share the [`temporary timeout fix`](https://github.com/github/github/pull/272079/files#diff-9001bfb28bbac746e708f1fc171f96bcfac349b6ecc94adb7ec80a7da167cf98R803-R808) fix publicly?
> 
> Thanks!
> 
> **Edit:** In the meantime I found a css workaround for the issue. Not an ideal solution, but it does the trick. Hide the tooltip, when the element is focused, not hovered and does not have the focus visible. This means to hide the tooltip when the focus is not triggered via a pointer or the keyboard.
> 
> ```css
> button:focus:not(:focus-visible):not(:hover) + tool-tip {
>   visibility: hidden;
> }
> ```

Fyi: For us, the issue was fixed with version 0.16.0 due to [this commit](https://github.com/primer/view_components/commit/87e03ea4dd41822454c72ddb7e55b0aa9294c69c). 🥳

**Comment by TylerJDev** [comment]
@randall-krauskopf, I'm taking another look at this issue, and I'm wondering if there's anyway to create a similar environment to what you've shown in your video examples? I see you've mentioned that the addition of a timeout does not work for hotkeys specifically. Also, if we're using a custom function to handle the hotkeys, could you share where this lives? Thanks too :grin:

**Comment by randall-krauskopf** [comment]
@TylerJDev We removed the hotkey functionality for accessibility purposes I believe so I don't have a similar environment ready to go atm. Also, trying to remember what I did 7 months ago is a bit hazy 😄 .

**Comment by TylerJDev** [comment]
No worries @randall-krauskopf! I do realize that there might be some accessibility implications here, since we're dealing with hotkeys in the menu. Since this is no longer an active problem, would it be okay if I closed this issue for now? 

If we want to introduce hotkey/focus management functionality in `ActionMenu` in the future then we can reopen this issue, or create a new one in [Primer](https://github.com/github/primer).

**Comment by randall-krauskopf** [comment]
I think you can go ahead and close it.

**Comment by TylerJDev** [comment]
Closing this issue for now. If we want to revisit this topic then we can reopen the issue, or create a new one!
